src/Nethermind/Nethermind.AuRa/Validators/ContractValidator.cs

```diff
@@ -65,68 +83,62 @@ public class ContractValidator : IAuRaValidatorProcessor
                 throw new ArgumentException("Wrong validator type.", nameof(validator));
             
             ContractAddress = validator.Addresses?.FirstOrDefault() ?? throw new ArgumentException("Missing contract address for AuRa validator.", nameof(validator.Addresses));
+            _stateDb = stateDb ?? throw new ArgumentNullException(nameof(stateDb));
             _stateProvider = stateProvider ?? throw new ArgumentNullException(nameof(stateProvider));
             _transactionProcessor = transactionProcessor ?? throw new ArgumentNullException(nameof(transactionProcessor));
             _logger = logManager?.GetClassLogger() ?? throw new ArgumentNullException(nameof(logManager));
             AbiEncoder = abiEncoder ?? throw new ArgumentNullException(nameof(abiEncoder));
-            StartBlockNumber = startBlockNumber;
+            InitBlockNumber = startBlockNumber;
+            LoadPendingValidators();
         }
-        
+
         public bool IsValidSealer(Address address) => _validators.Contains(address);
+        
+        public int MinSealersForFinalization => _validators.MinSealersForFinalization();
 
-        public virtual void PreProcess(Block block)
+        void IAuRaValidator.SetFinalizationManager(IBlockFinalizationManager finalizationManager)
         {
```

Comment:
in what cases this would be set twice?

Reply:
should it not assume only single setup and throw exception on consecutive attempts?